composer/utils/batch_helpers.py

```diff
@@ -100,37 +71,44 @@ def batch_set(batch: Any,
     ...     batch[1]['d'] = value
     ...     return batch
     ...
-    >>> batch_set([{'a':1, 'b':7},{'d':3}], value=20, set_fn=setter)
+    >>> batch_set([{'a':1, 'b':7},{'d':3}], key=setter, value=20)
     [{'a': 1, 'b': 7}, {'d': 20}]
+    >>> batch_set([{'a':1, 'b':7},{'d':3}], key=(lambda x: x[0]['b'], setter), value=20)
+    [{'a': 1, 'b': 7}, {'d': 20}]
+
 
     Args:
         batch (Any): An object that contains the input and label of the items in the batch.
             Can be any abritrary type that user creates, but we assume some sort of
             sequence (list, tuple, tensor, array), mapping (dictionary),
             or attribute store (object with data members, namedtuple).
-        key (Any): A key to index into the batch.
+        key (str, int, or Callable): A key to index into the batch or a user-specified function 
+            to do the setting. A pair of callables is also supported for cases where a get
+            and set function pair are both passed (like in Algorithms). The setter is
+            assumed to be the second of the pair.
         value (Any): The value that batch[key] or batch.key gets set to.
-        set_fn (Callable): A user-specified function to do the setting. set_fn is optional if key and 
-            value are supplied. set_fn must return the updated batch.
 
     Returns:
         batch (Any): updated batch with value set at key.
 
-    Raises:
-        ValueError if:
-            * key and set_fn are both unset
-            * key and set_fn are both set
     """
-    if key is None and set_fn is None:
-        raise ValueError("key or set_fn must be specified and neither were!")
-    if key is not None and set_fn is not None:
-        raise ValueError("key and set_fn were both set. Only one can be set!")
-    if set_fn:
-        return set_fn(batch, value)
+    batch_copy = copy.deepcopy(batch)
```

Comment:
Curious why do we copy the batch? I think I would expect that if the batch is a mutable object, then I would want it to be modified in place. (Also not sure whether this would work for GPU tensors -- could cause a memory leak maybe?)

Reply:
It shouldn't cause memory leak because I think the reference count would go to zero in this case.

Reply:
But you're right I don't think we need the deepcopy. Fixed here: [a45b966](https://github.com/mosaicml/composer/pull/1041/commits/a45b9669192efabd3f2607520cb5b0ab3d346a71)